services/pull/protection_check.go

```diff
@@ -0,0 +1,86 @@
+// Copyright 2022 The Gitea Authors. All rights reserved.
+// Use of this source code is governed by a MIT-style
+// license that can be found in the LICENSE file.
+
+package pull
+
+import (
+	"context"
+
+	"code.gitea.io/gitea/models"
+	user_model "code.gitea.io/gitea/models/user"
+	asymkey_service "code.gitea.io/gitea/services/asymkey"
+)
+
+// CheckPullProtection check if the pull mergable based on all conditions (branch protection, merge options, ...)
+func CheckPullProtection(ctx context.Context, doer *user_model.User, perm *models.Permission, pr *models.PullRequest, manuallMerge, force bool) error {
+	if pr.HasMerged {
+		return ErrHasMerged{}
+	}
+
+	if err := pr.LoadIssue(); err != nil {
+		return err
+	} else if pr.Issue.IsClosed {
+		return ErrIsClosed{}
+	}
+
+	if allowedMerge, err := IsUserAllowedToMerge(pr, *perm, doer); err != nil {
+		return err
+	} else if !allowedMerge {
+		return ErrUserNotAllowedToMerge{}
+	}
+
+	if manuallMerge {
+		// don't check rules to "auto merge", doer is going to mark this pull as merged manually
+		return nil
+	}
+
+	if pr.IsWorkInProgress() {
+		return ErrIsWorkInProgress{}
+	}
+
+	if !pr.CanAutoMerge() {
+		return ErrNotMergableState{}
+	}
+
+	if err := CheckPRReadyToMerge(ctx, pr, false); err != nil {
+		if models.IsErrNotAllowedToMerge(err) {
```

Comment:
I think we can choose a new name for `NotAllowedToMerge`, maybe `DisallowedToMerge`

`IsErrNotAllowedToMerge` reads a little strange. And it would be more strange to understand `!IsErrNotAllowedToMerge()` 😂